src/shared/lib/FisherExactTestCalculator.ts

```diff
@@ -0,0 +1,35 @@
+function getPValue(a: number, b: number, c: number, d: number, f: number[]): number {
+
+    const n: number = a + b + c + d;
+    const p: number = (f[a + b] + f[c + d] + f[a + c] + f[b + d]) - (f[a] + f[b] + f[c] + f[d] + f[n]);
+    return Math.exp(p);
+}
+
+export function getCumulativePValue(a: number, b: number, c: number, d: number): number {
+
+    let min: number, i: number;
+    const n: number = a + b + c + d;
+    let p: number = 0;
+    let f: number[] = new Array(n + 1);
+    f[0] = 0.0;
+
+    for (let j: number = 1; j <= n; j++) {
+        f[j] = f[j - 1] + Math.log(j);
+    }
+
+    p += getPValue(a, b, c, d, f);
+    if ((a * d) >= (b * c)) {
+        min = (c < b) ? c : b;
+        for (i = 0; i < min; i++) {
+            p += getPValue(++a, --b, --c, ++d, f);
+        }
+    }
+
+    if ((a * d) < (b * c)) {
+        min = (a < d) ? a : d;
+        for (i = 0; i < min; i++) {
+            p += getPValue(--a, ++b, ++c, --d, f);
+        }
+    }
```

Comment:
I haven't thought through the math - does this strategy add all of, and only, the probabilities that are <= prob(a,b,c,d)?